tests/unit/Core/Framework/DataAbstractionLayer/ExtensionRegistryTest.php

```diff
@@ -0,0 +1,85 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Tests\Unit\Core\Framework\DataAbstractionLayer;
+
+use PHPUnit\Framework\Attributes\CoversClass;
+use PHPUnit\Framework\TestCase;
+use Shopware\Core\Content\Category\CategoryDefinition;
+use Shopware\Core\Content\Product\ProductDefinition;
+use Shopware\Core\Framework\DataAbstractionLayer\BulkEntityExtension;
+use Shopware\Core\Framework\DataAbstractionLayer\Dbal\EntityWriteGateway;
+use Shopware\Core\Framework\DataAbstractionLayer\EntityDefinition;
+use Shopware\Core\Framework\DataAbstractionLayer\ExtensionRegistry;
+use Shopware\Core\Framework\DataAbstractionLayer\Field\FkField;
+use Shopware\Core\Framework\DataAbstractionLayer\Field\ManyToOneAssociationField;
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\Test\Stub\DataAbstractionLayer\StaticDefinitionInstanceRegistry;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+/**
+ * @internal
+ */
+#[Package('core')]
+#[CoversClass(ExtensionRegistry::class)]
+class ExtensionRegistryTest extends TestCase
+{
+    public function testNoBulkRegistered(): void
+    {
+        $registry = new ExtensionRegistry([], []);
+
+        $bulks = $registry->buildBulkExtensions(
+            $this->registry([])
+        );
+
+        static::assertEmpty($bulks);
+    }
+
+    public function testUnkownDefinition(): void
+    {
+        $registry = new ExtensionRegistry([], [new MyBulkExtension()]);
+
+        $bulks = $registry->buildBulkExtensions(
+            $this->registry([ProductDefinition::class])
+        );
+
+        static::assertCount(1, $bulks);
+    }
+
+    public function testAllDefinitionsFound(): void
+    {
+        $registry = new ExtensionRegistry([], [new MyBulkExtension()]);
+
+        $bulks = $registry->buildBulkExtensions(
+            $this->registry([ProductDefinition::class, CategoryDefinition::class])
+        );
+
+        static::assertCount(2, $bulks);
+    }
+
+    /**
+     * @param array<int|string, class-string<EntityDefinition>|EntityDefinition> $definitions
+     */
+    private function registry(array $definitions): StaticDefinitionInstanceRegistry
+    {
+        return new StaticDefinitionInstanceRegistry(
+            $definitions,
+            $this->createMock(ValidatorInterface::class),
+            $this->createMock(EntityWriteGateway::class)
+        );
+    }
+}
+
+class MyBulkExtension extends BulkEntityExtension
+{
+    public function collect(): \Generator
+    {
+        yield 'product' => [
```

Comment:
looking at the docs and the code here now in the registry I think it would be better to use the definition class here as key or is there a benefit to use the entity name here?

With
```
yield ProductDefinition::class => [...]
```

There would be no need to pass the definition registry into the extension registry and that would match how you declare "normal" extensions as well 

What do you think?

Reply:
I think the benefit of using the entity name is compatibility with entities via attributes. Am I right? https://developer.shopware.com/docs/guides/plugins/plugins/framework/data-handling/entities-via-attributes.html

Reply:
With the attribute system and custom entities, the definition classes are no more required. Therefore nearly everywhere the dal has to switch to entity names instead of the class definitions

Reply:
BTW i guess we discovered a new test case that would be important to cover

The longer I think about it the more I have the impression that all the compiling should be moved from the framework boot to the extension registry 
so all that is left in framework should be 
```
$registry->registerExtensions($definitionRegistry, $salesChannelDefinitionRegistry);
```

That way we could cover all of the connecting logic with tests

For reference see https://martinfowler.com/bliki/TellDontAsk.html

Reply:
Okay, I understand. I would actually approach it the other way around. The use of definitions has become more optional over the years, and recent refactorings have increasingly moved towards the DAL working more with entity names (to support attribute entities and custom entities in apps) rather than definitions.

- Therefore, I would let the bulk entity extensions work directly with entity names.
- Add support for "entity name" in the existing entity extension.
- Shift the extension system to focus on "entity name"

Would this be okay?

Reply:
Would be fine when the extensions also support the entity name instead of definition then, and probably directly deprecate the definition way of doing this

Reply:
By the way i guess this comment from @JoshuaBehrens  is also super relevant here: https://github.com/shopware/docs/pull/1543#discussion_r1833212880

@JoshuaBehrens would this go in the right direction for you as well? Or what additional issues did you encounter?

Reply:
I wrote him an answer, IMO i don't like this "i need a constant - i dont want to rely on a string which can be changed" in case of entities. 

Yes a table name could change, but shopware is years away IMO. I would always prefer the way to use the table name instead of a constant, which makes it much better readable.

But like i wrote at the docs, for all which wants a constant, i would recommand to re-add a class Constant for the entity name

Reply:
I also like in general constants, but not that I want to change them, but that they autocomplete in the IDE :-)

I do not want to change the content of them (as you already mentioned, it will and should probably never happen, as it is quite a break), but not having them autocomplete, introduces a source of bugs which can be avoided and one always needs to think about if one properly spelled the table name.